### ISSUE TITLE
Add preloading support to CMake and other build commands

### ISSUE THREAD

**Issue by rui314** [rui314/mold#74]
mold supports object file preloading feature. In order to use it, someone has to run a linker command with `--preload` flag a few seconds prior to the actual linker invocation. And I think "someone" should be a build system, such as make or ninja executing CMake-generated ninja or makefiles.

As a starter, I think we should add the following feature to CMake:

- Detect if a linker is mold
- run a linker command with `--preload` if it is mold before invoking compiler instances to generate object files

**Comment by comicfans** [comment]
now I'm using CMake to generate ninja build, and using mold -run ninja to build whole project, 
mold performance is impressive during my test, I'd like to know if mold can implement
```
mold --preload -run ninja
```
it doesn't touch cmake, and mold is known to be running at first.

**Comment by rui314** [comment]
@comicfans I think such feature cannot be implemented. In order to do preloading, mold has to know what ninja will do next, but it is generally not predictable.

**Comment by comicfans** [comment]
so if I understand correctly, this feature expected to run mold daemon (with exactly same link  arguments) even before all input object compile complete, and for real link, it already preload most input?  previously I misunderstood preload as 'single daemon to cache all following invocation'

**Comment by rui314** [comment]
Yes, the daemon is for single-use. In order to use the preloading feature, you invoke mold twice instead of once for each linker output. The first one is invoked with `--preload` to preload object files, and the second one is invoked without `--preload` to tell the daemon to finish its job.

**Comment by gruenich** [comment]
I opened an issue in CMake's GitLab, see https://gitlab.kitware.com/cmake/cmake/-/issues/23063

**Comment by daquexian** [comment]
It looks not a hard task. Maybe I can implement it in CMake (if I have time). I noticed that preload code has already been removed from the latest mold. I can firstly try with some old version mold. If it succeeds, will you re-introduce the preload feature into master?  @rui314

**Comment by rui314** [comment]
I removed `--preload` because there's no use of it. If it's supported by a build system and proved to be useful, we can discuss resurrecting it.

If you use `--preload`, you invoke the linker twice with and without `--preload`. You first invoke the linker with `--preload` and after a few seconds, run the same command without `--preload`. If you always run mold with `--preload`, it doesn't do anything but just preloading files (so no output would be created).

**Comment by daquexian** [comment]
> If you use --preload, you invoke the linker twice with and without --preload. You first invoke the linker with --preload and after a few seconds, run the same command without --preload. If you always run mold with --preload, it doesn't do anything but just preloading files (so no output would be created).

That is actually what I did.

Here is the command (the libraries I want to generate is `liboneflow.so`, and the `time` command output is zsh format):

```bash
$ rm liboneflow.so
$ time ./link.sh
./tmp.sh  0.06s user 0.03s system 2% cpu 3.124 total
$ rm liboneflow.so
$ ./link_preload.sh
# after cpu becomes idle:
$ time ./link.sh
./tmp.sh  0.05s user 0.05s system 2% cpu 3.294 total
```

**Comment by rui314** [comment]
0.06s is already too fast, so I don't think you can observe any improvements over it. I think it needs to take at least a few seconds to see an improvement.

**Comment by daquexian** [comment]
> 0.06s is already too fast, so I don't think you can observe any improvements over it. I think it needs to take at least a few seconds to see an improvement.

It is zsh format (which is [different](https://unix.stackexchange.com/questions/562650/time-command-is-different-in-zsh-than-in-bash) with bash) so I think the "total" item (3.124 and 3.294) is the wall clock time.

**Comment by rui314** [comment]
@daquexian You can add `--perf` to see the breakdown of internal passes. Can you try that option?

**Comment by daquexian** [comment]
What I tried:

```bash
$ ./link.sh > no_preload
$ ./link_preload.sh
$ ./link.sh > preload
$ ld --version
mold 1.0.0 (ed9924895d9b9584106791247596677db8113528; compatible with GNU ld and GNU gold)
```

I found there is still a 1-second "copy" in `preload`.

I uploaded the script I used and the `--perf` output. Download link:
[link.sh](https://oshi.at/SSqY/link.sh)
[link_preload.sh](https://oshi.at/apgu/link_preload.sh)
[preload](https://oshi.at/xkYh/preload)
[no_preload](https://oshi.at/UXtf/no_preload)

**Comment by daquexian** [comment]
> I don't know why, but it looks like preloading didn't work at all for your test case. It might be a bug.

I see. I'll try to support preload in cmake (when I have time) anyway.
